src/uct/ib/mlx5/gdaki/gdaki.c

```diff
@@ -42,48 +49,169 @@ static UCS_CLASS_INIT_FUNC(uct_rc_gdaki_ep_t, const uct_ep_params_t *params)
     uct_ib_mlx5_md_t *md = ucs_derived_of(iface->super.super.super.super.md,
                                           uct_ib_mlx5_md_t);
     uct_ib_iface_init_attr_t init_attr = {};
+    uct_ib_mlx5_cq_attr_t cq_attr      = {};
     uct_ib_mlx5_qp_attr_t qp_attr      = {};
+    uct_rc_gdaki_dev_ep_t dev_ep       = {};
     ucs_status_t status;
+    doca_error_t derr;
+    cudaError_t cerr;
+    size_t dev_ep_size;
+    uct_ib_mlx5_dbrec_t dbrec;
 
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super.super.super.super);
 
-    init_attr.cq_len[UCT_IB_DIR_TX] = iface->super.super.config.tx_qp_len *
-                                      UCT_IB_MLX5_MAX_BB;
-    init_attr.flags                 = UCT_IB_CQ_IGNORE_OVERRUN;
-    status = uct_ib_mlx5_devx_create_cq(&iface->super.super.super,
-                                        UCT_IB_DIR_TX, &init_attr, &self->cq, 0,
-                                        0);
+    status = uct_cuda_copy_push_ctx(iface->cuda_dev, 1, UCS_LOG_LEVEL_ERROR);
     if (status != UCS_OK) {
         return status;
     }
 
+    init_attr.cq_len[UCT_IB_DIR_TX] = iface->super.super.config.tx_qp_len *
+                                      UCT_IB_MLX5_MAX_BB;
+    uct_ib_mlx5_cq_calc_sizes(&iface->super.super.super, UCT_IB_DIR_TX,
+                              &init_attr, 0, &cq_attr);
     uct_rc_iface_fill_attr(&iface->super.super, &qp_attr.super,
                            iface->super.super.config.tx_qp_len, NULL);
+    uct_ib_mlx5_wq_calc_sizes(&qp_attr);
+
+    cq_attr.flags      |= UCT_IB_MLX5_CQ_IGNORE_OVERRUN;
+    cq_attr.umem_offset = ucs_align_up_pow2(
+            sizeof(uct_rc_gdaki_dev_ep_t) +
+                    qp_attr.max_tx * sizeof(uct_rc_gdaki_op_t),
+            ucs_get_page_size());
+
     qp_attr.mmio_mode     = UCT_IB_MLX5_MMIO_MODE_DB;
     qp_attr.super.srq_num = 0;
-    status = uct_ib_mlx5_devx_create_qp(&iface->super.super.super, &self->cq,
-                                        &self->cq, &self->qp.super, &self->qp,
-                                        &qp_attr);
+    qp_attr.umem_offset   = ucs_align_up_pow2(cq_attr.umem_offset +
+                                                      cq_attr.umem_len,
+                                              ucs_get_page_size());
+
+    dev_ep_size = qp_attr.umem_offset + qp_attr.len;
```

Comment:
1. do we need space for control structure like qp/cq indexes?
2. maybe add short ascii drawing to describe the layout